[DO NOT MERGE] feat: Camel K 2.0 release notes (#1032)
apache/camel-website: content/blog/2023/07/camel-k-2-0/index.md

This is a good news, because it will free the operator from the burden of such an heavy operation. Moreover we introduced some sensible resource configuration which should be enough for the builder Pod to complete its work without affecting the stability of the operator.

However, if those are not enough or you still want to customize the builder Pod resources, we have made available a set of [builder parameters](/camel-k/next/traits/builder.html) (mainly cpu and memory resources).

## Run build unprivileged

This is quite an hidden work we've done and it really gave us a lot of trouble in order to let **Spectrum** and **S2I** builds to run in an unprivileged mode. However, we've accomplished it and we feel a little safer against any possible attack from privileged escalation.

NOTE: we haven't (yet) deprecated **Kaniko** and **Buildah** strategy and these deployment methods require to run with `root` privileges. Make sure to understand the possible risks involved.

> ## Build order strategy

Reviewer: ```suggestion
## Build order strategy

Another interesting feature that is now available to you is the possibility to define a build order strategy that controls in which order scheduled integration builds get executed. 

In Camel K v1 a series of parallel submitted builds got executed strictly sequential (only one single build at a time) and there was no guarantee on the order of build execution. 

Now you can specify the order of builds using the [Builder trait order strategy parameter](/camel-k/next/traits/builder.html#_configuration).

Supported build order strategies are:
- `sequential`: equivalent to the Camel K 1 behavior running only one single build at a time
- `fifo`: First in first out strategy runs builds as they have been created and allows builds to run in parallel to each other as long as `maxRunningBuilds` limit has not been reached.
- `dependencies`: Sophisticated order strategy smartly creating a queue for those builds which may depend on each other in order to effectively run builds leveraging the incremental build feature. Allows independent builds to run in parallel to each other as long as `maxRunningBuilds` limit has not been reached.

You can now choose from these build order strategies to get the best out of Camel K. Our operator is smarter than any AI out there!!
```

> ## Build order strategy
> 
> Another little interesting feature you have available now is the possibility to define which is the order of the building strategy. When in
> Camel K 1 you submitted a series of parallel builds, there was no guarantee on the order of execution. Now you can specify it using the [Builder trait order strategy parameter](/camel-k/next/traits/builder.html#_configuration).
> 
> If you run many builds in parallel you may want to use the `dependencies` strategy which will take care to smartly create a queue for those builds which may depend on each other and leverage the incremental build feature.
> 
> Our operator is smarter than any AI out there!!

Reviewer: ```suggestion
## Build order strategy

Another interesting feature that is now available to you is the possibility to define a build order strategy that controls in which order scheduled integration builds get executed. 

In Camel K v1 a series of parallel submitted builds got executed strictly sequential (only one single build at a time) and there was no guarantee on the order of build execution.

Now you can specify the order of builds using the [Builder trait order strategy parameter](/camel-k/next/traits/builder.html#_configuration).

Supported build order strategies are:
- `sequential`: equivalent to the Camel K 1 behavior running only one single build at a time
- `fifo`: First in first out strategy runs builds as they have been created and allows builds to run in parallel to each other as long as `maxRunningBuilds` limit has not been reached.
- `dependencies`: Sophisticated order strategy smartly creating a queue for those builds which may depend on each other in order to effectively run builds leveraging the incremental build feature. Allows independent builds to run in parallel to each other as long as `maxRunningBuilds` limit has not been reached.
You can now choose from these build order strategies to get the best out of Camel K. Our operator is smarter than any AI out there!!
```

> 

Reviewer: ```suggestion
## Build order strategy
Another interesting feature that is now available to you is the possibility to define a build order strategy that controls in which order scheduled integration builds get executed. 
In Camel K v1 a series of parallel submitted builds got executed strictly sequential (only one single build at a time) and there was no guarantee on the order of build execution. 
Now you can specify the order of builds using the [Builder trait order strategy parameter](/camel-k/next/traits/builder.html#_configuration).
Supported build order strategies are:
- `sequential`: equivalent to the Camel K 1 behavior running only one single build at a time
- `fifo`: First in first out strategy runs builds as they have been created and allows builds to run in parallel to each other as long as `maxRunningBuilds` limit has not been reached.
- `dependencies`: Sophisticated order strategy smartly creating a queue for those builds which may depend on each other in order to effectively run builds leveraging the incremental build feature. Allows independent builds to run in parallel to each other as long as `maxRunningBuilds` limit has not been reached.
You can now choose from these build order strategies to get the best out of Camel K. Our operator is smarter than any AI out there!!
```

This change won't affect directly the users of Camel K but it's important to mention for those tools or application that are using **Camel K Golang API**. We have upgraded to V2 accordingly, so, anybody willing to use this API will have to do the update accordingly.

We have also performed a deep review to verify those old unsupported dependencies which may lead to a possible headache in case of some CVE. We have removed all the unsupported dependencies and we feel a bit more **safer against possible vulnerabilities** in the dependencies we use.

# Remove deprecated CLI commands

As we embarked in a major upgrade, we took the opportunity to clean all the **deprecated CLI commands** and configuration we had marked in Camel K version 1.

* `kamel install --kaniko-build-cache`
* `kamel local`
* `kamel init`
* `kamel run --config|--resource file:`

In version 2 they won't be any longer available and the operator may fail if those configurations are provided. Above all, the last one will require you to provide a resource file as either a Configmap or Secret if you want to use the related configuration (we don't support any longer the auto-generated Confimaps from file).

> # Multiple data types and schemas in Kamelets

Reviewer: ```suggestion
# Multiple data types and schemas in Kamelets

Kamelets are now able to expose supported input/output data types as part of the [Kamelet data type specification](/camel-k/next/kamelets/kamelets-dev.html#_kamelet_data_types) in order to increase the interoperability with other components. When referencing a Kamelet as a source or sink in a Pipe users may now choose an appropriate input/output data type that fits best for their individual use case. 

This feature leverages the [data types transformer concept](https://camel.apache.org/manual/transformer.html) in Apache Camel 4 and reveals its full power once the new major version of Camel is available (we are all eagerly awaiting it).
```

This feature is not immediately available as it depends on the presence of the relative feature which will be available in Camel 4. When a Camel 4 compatible runtime will be available (we're all waiting with impatience), then you will be able to use the new [Kamelet Data Type configuration](https://github.com/apache/camel-k/pull/4100).

> TODO - include some documentation instead of the link to the PR.

Reviewer: ```suggestion
# Multiple data types and schemas in Kamelets

Kamelets are now able to expose supported input/output data types as part of the [Kamelet data type specification](/camel-k/next/kamelets/kamelets-dev.html#_kamelet_data_types) in order to increase the interoperability with other components. When referencing a Kamelet as a source or sink in a Pipe users may now choose an appropriate input/output data type that fits best for their individual use case. 

This feature leverages the [data types transformer concept](https://camel.apache.org/manual/transformer.html) in Apache Camel 4 and reveals its full power once the new major version of Camel is available (we are all eagerly awaiting it).
```

# Multi architecture support

This feature should not be considered as generally available but more a kind of "tech preview". In Camel K 2.0 we provide an **ARM64 architecture based container image** for the operator but it is not tested due to the difficulty to find the required infrastructure and have the suite of test running. It's a best effort feature that we hope we can make generally available in the future versions.

For more information you can have a look at the [Multi architecture official documentation](/camel-k/next/installation/advanced/multi-architecture.html) page.

# Code coverage report

If you're a Camel K contributor you have noticed that we introduced a nice action which is taking care to calculate the code coverage of the new code developed. The actions reports the coverage difference, so we know how we increase (or decrease) indirectly the quality of the code. This is not really (yet) a gateway, we want to create awareness among Camel K developers about the level of quality we introduce with each change. Little by little we expect to **increase the quality of the code** we produce and have less bugs affecting our code.

> 

Reviewer: ```suggestion
# Optimize operator performance and resource allocation

The team has made some effort to reduce the amount of resources (memory and CPU) consumed by the Camel K operator. The results of this can be seen when creating many Integrations on a cluster and when the operator runs multiple integration builds.

We have run some test scenarios and compared the results of different Camel K versions.

_Test scenario: Running 600 Camel K integrations_

| Camel K | avg. Memory | avg. CPU | Time |
| :------- | :------------: | :--------: | :----: |
| 1.9.0 | 529 MiB  | 262 millicores | 29 mins |
| 1.12.1 | 344 MiB | 54 millicores | 21 mins |
| 2.0.0 | 306 MiB | 60 millicores | 20 mins |

_Test scenario: Running 50 integrations builds in parallel_

| Camel K | avg. Memory | avg. CPU | Time |
| :------- | :------------: | :--------: | :----: |
| 1.9.0 | 510 MiB | 352 millicores | 59 mins |
| 1.12.1 | 440 MiB | 185 millicores | 57 mins |
| 2.0.0 | 326 MiB | 45 millicores | 38 mins |

The Camel K 2.0 results are pleasant especially when running multiple builds at the same time. The team is keen to keep improving the operator performance in future releases.
```
